certora/specs/RevertsMintBurnEthereum.spec

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+methods {
+    function owner() external returns address envfree;
+    function totalSupply() external returns uint256 envfree;
+    function balanceOf(address) external returns uint256 envfree;
+    function delegatee(address) external returns address envfree;
+    function delegatedVotingPower(address) external returns uint256 envfree;
+}
+
+// Check the revert conditions for the burn function.
+rule mintRevertConditions(env e, address to, uint256 amount) {
+    mathint totalSupplyBefore = totalSupply();
+    uint256 balanceOfSenderBefore = balanceOf(e.msg.sender);
+    uint256 toVotingPowerBefore = delegatedVotingPower(delegatee(to));
+
+    // Safe require as zero address can't possibly delegate voting power.
+    require delegatee(0) == 0;
+
+    // Safe require because if the delegatee is not zero the recipient's voting power excludes the newly minted value.
+    require delegatee(to) != 0 => toVotingPowerBefore <= totalSupply() - amount;
+
+    mint@withrevert(e, to, amount);
+    assert lastReverted <=> e.msg.sender != owner() || to == 0 || e.msg.value != 0 || totalSupplyBefore + amount > max_uint256;
+}
+
+// Check the revert conditions for the burn function.
+rule burnRevertConditions(env e, uint256 amount) {
+    uint256 balanceOfSenderBefore = balanceOf(e.msg.sender);
+    uint256 senderVotingPowerBefore = delegatedVotingPower(delegatee(e.msg.sender));
+
+    // Safe require as zero address can't possibly delegate voting power.
+    require delegatee(0) == 0;
```

Comment:
Indeed, actually it's proven here https://github.com/morpho-org/morpho-token-upgradeable/blob/04ecb92c4a08e6801d8379cc0465e75556dae860/certora/specs/Delegation.spec#L14. I just can't import it just yet because it's in another branch :(

Reply:
I think there would be quite many conflicts (related to CI and the README mainly). Instead, this one should be merged after the delegation PR.

Reply:
@QGarchery there is a downside related to this. Notably, using the invariant mentioned in this https://github.com/morpho-org/morpho-token/pull/98#discussion_r1853924160, this also requires to run the verification on the munged files which catches revert conditions related to overflows when updated the virtual delegated power to address zero. I've implemented this in https://github.com/morpho-org/morpho-token/pull/98/commits/94fef53c70387af12e07e830685f67c5fdd127cf to give it a try. Let me know what you think.

Reply:
Hm annoying, ok let's revert back

Reply:
Instead we can point to the invariant that we need. So instead of doing:
```solidity
// Safe require because propertyA holds.
require propertyA;
```
I think we should do:
```solidity
// Safe require because it is verified in invariantXXX.
require propertyA;
```